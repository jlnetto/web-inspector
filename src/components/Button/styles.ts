import styled, { keyframes } from 'styled-components';

export const StyledButton = styled.button`
  min-width: 100px;
  padding: 10px 20px;
  font-size: 14px;
  background-color: #007bff;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;

  &:hover {
    background-color: #0056b3;
  }

  &:disabled {
    background-color: #808080;
    cursor: not-allowed;
    opacity: 0.7;
  }
`;

const rotate360 = keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
`;

export const Spinner = styled.div`
  animation: ${rotate360} 1s linear infinite;
  transform: translateZ(0);

  border-top: 2px solid #808080;
  border-right: 2px solid #808080;
  border-bottom: 2px solid #808080;
  border-left: 4px solid white;
  background: transparent;
  width: 15px;
  height: 15px;
  border-radius: 50%;
`;
