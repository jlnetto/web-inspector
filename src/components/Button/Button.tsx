import { StyledButton, Spinner } from './styles';

export type ButtonProps = {
  type: 'submit' | 'button';
  text: string;
  onClick?: () => void;
  disabled?: boolean;
  loading?: boolean;
};

const Button: React.FC<ButtonProps> = ({
  text,
  disabled,
  loading,
  ...props
}) => (
  <StyledButton {...props} disabled={disabled || loading}>
    {loading ? <Spinner data-testid="spinner" /> : text}
  </StyledButton>
);

export default Button;
