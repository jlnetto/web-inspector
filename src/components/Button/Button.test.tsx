import { render, fireEvent, screen } from '@testing-library/react';
import Button, { ButtonProps } from './Button';

describe('Button', () => {
  const handleClick = jest.fn();

  beforeEach(() => {
    handleClick.mockClear();
  });

  it('renders with text', () => {
    const props: ButtonProps = {
      text: 'Click me',
      onClick: handleClick,
      type: 'button',
    };
    const { asFragment } = render(<Button {...props} />);

    expect(screen.getByText('Click me')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with default button type', () => {
    const props: ButtonProps = {
      text: 'Click me',
      onClick: handleClick,
      type: 'button',
    };
    render(<Button {...props} />);
    expect(screen.getByText('Click me')).toHaveAttribute('type', 'button');
  });

  it('renders with submit button type', () => {
    const props: ButtonProps = {
      text: 'Submit',
      onClick: handleClick,
      type: 'submit',
    };
    render(<Button {...props} />);
    expect(screen.getByText('Submit')).toHaveAttribute('type', 'submit');
  });

  it('calls onClick handler when clicked', () => {
    const props: ButtonProps = {
      text: 'Click me',
      onClick: handleClick,
      type: 'button',
    };
    render(<Button {...props} />);
    fireEvent.click(screen.getByText('Click me'));
    expect(handleClick).toHaveBeenCalled();
  });

  it('does not call onClick handler when disabled', () => {
    const props: ButtonProps = {
      text: 'Click me',
      onClick: handleClick,
      disabled: true,
      type: 'button',
    };
    render(<Button {...props} />);
    fireEvent.click(screen.getByText('Click me'));
    expect(handleClick).not.toHaveBeenCalled();
  });

  it('render with loading spinner when loading is active', () => {
    const props: ButtonProps = {
      text: 'Click me',
      onClick: handleClick,
      disabled: true,
      type: 'button',
      loading: true,
    };
    render(<Button {...props} />);

    expect(screen.getByTestId("spinner")).toBeInTheDocument();
  });
});
