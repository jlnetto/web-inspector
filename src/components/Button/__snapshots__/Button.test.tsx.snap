// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders with text 1`] = `
<DocumentFragment>
  <button
    class="sc-bgqQcB kYYpOZ"
    type="button"
  >
    Click me
  </button>
</DocumentFragment>
`;
